lgc/patch/PatchFetchShader.cpp

```diff
@@ -0,0 +1,294 @@
+ */
+#include "SystemValues.h"
+#include "VertexFetch.h"
+#include "lgc/LgcContext.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PipelineShaders.h"
+#include "lgc/state/PipelineState.h"
+#include "lgc/state/ResourceUsage.h"
+#include "lgc/state/TargetInfo.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "llpc-patch-copy-shader"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Pass to generate fetch shader if required
+class PatchFetchShader : public Patch {
+public:
+  static char ID;
+  PatchFetchShader() : Patch(ID) {}
+
+  bool runOnModule(Module &module) override;
+
+  void getAnalysisUsage(AnalysisUsage &analysisUsage) const override {
+    analysisUsage.addRequired<PipelineStateWrapper>();
+    analysisUsage.addRequired<PipelineShaders>();
+    // Pass does not preserve PipelineShaders as it adds a new shader.
+  }
+
+private:
+  PatchFetchShader(const PatchFetchShader &) = delete;
+  PatchFetchShader &operator=(const PatchFetchShader &) = delete;
+
+  FunctionType *generateFetchShaderEntryPointType(uint64_t *inRegMask);
+  Function *createEntryPoint();
+  void buildFetchShaderBody(Function *entryPoint);
+  Value *fetchVertexInput(Type *fetchType, Type *returnTy, unsigned location, Instruction *insertPos,
+                          VertexFetch *vertexFetch);
+
+  PipelineState *m_pipelineState;               // Pipeline state
+  std::unique_ptr<llvm::IRBuilder<>> m_builder; // The IRBuilder.
+};
+
+char PatchFetchShader::ID = 0;
+
+} // namespace lgc
+
+// =====================================================================================================================
+// Create pass to generate copy shader if required.
+ModulePass *lgc::createPatchFetchShader() {
+  return new PatchFetchShader();
+}
+
+// =====================================================================================================================
+// Run the pass on the specified LLVM module.
+bool PatchFetchShader::runOnModule(Module &module) // [in,out] LLVM module to be run on
+{
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Fetch-Shader\n");
+  Patch::init(&module);
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+  m_builder.reset(new IRBuilder<>(module.getContext()));
+
+  if (!m_pipelineState->getLgcContext()->buildingFetchShader()) {
+    return false;
+  }
+
+  Function *entryPoint = createEntryPoint();
+  buildFetchShaderBody(entryPoint);
+
+  return true;
+}
+
+// =====================================================================================================================
+// Generate the body of a fetch shader in the current context.
+//
+// @param entryPoint [in]: The entry point to be modified.
+void PatchFetchShader::buildFetchShaderBody(Function *entryPoint) {
```

Comment:
Please change this method to use builder style consistently.

Reply:
Sorry, I don't know what you mean.  Can you be more specific?

Reply:
There's an IRBuilder<> in use, but I see a bunch of CastInst::CreateZExtOrBitCast, InsertValueInst::Create, ...; just use the corresponding m_builder->Create methods.

Reply:
Done